api/triage_metadata.go

```diff
@@ -0,0 +1,231 @@
+		return nil, err
+	}
+	newRef := &github.Reference{Ref: github.String("refs/heads/" + commitBranch), Object: &github.GitObject{SHA: baseRef.Object.SHA}}
+	ref, _, err = client.Git.CreateRef(tm.ctx, sourceOwner, sourceRepo, newRef)
+	return ref, err
+}
+
+// getTree generates the tree to commit based on the given files and the commit
+// of the ref you got in getRef.
+func (tm triageMetadata) getTree(ref *github.Reference, triagedMetadataMap map[string][]byte) (tree *github.Tree, err error) {
+	client := tm.githubClient
+	// Create a tree with what to commit.
+	entries := []github.TreeEntry{}
+
+	for folderPath, content := range triagedMetadataMap {
+		dest := shared.GetMetadataFilePath(folderPath)
+		entries = append(entries, github.TreeEntry{Path: github.String(dest), Type: github.String("blob"), Content: github.String(string(content)), Mode: github.String("100644")})
+	}
+
+	tree, _, err = client.Git.CreateTree(tm.ctx, sourceOwner, sourceRepo, *ref.Object.SHA, entries)
+	return tree, err
+}
+
+// createCommit creates the commit in the given reference using the given tree.
+func (tm triageMetadata) pushCommit(ref *github.Reference, tree *github.Tree) (err error) {
+	client := tm.githubClient
+	// Get the parent commit to attach the commit to.
+	parent, _, err := client.Repositories.GetCommit(tm.ctx, sourceOwner, sourceRepo, *ref.Object.SHA)
+	if err != nil {
+		return err
+	}
+	// This is not always populated, but is needed.
+	parent.Commit.SHA = parent.SHA
+
+	// Create the commit using the tree.
+	date := time.Now()
+	author := &github.CommitAuthor{Date: &date, Name: &tm.authorName, Email: &tm.authorEmail}
+	commit := &github.Commit{Author: author, Message: &commitMessage, Tree: tree, Parents: []github.Commit{*parent.Commit}}
+	newCommit, _, err := client.Git.CreateCommit(tm.ctx, sourceOwner, sourceRepo, commit)
+	if err != nil {
+		return err
+	}
+
+	// Attach the commit to the master branch.
+	ref.Object.SHA = newCommit.SHA
+	_, _, err = client.Git.UpdateRef(tm.ctx, sourceOwner, sourceRepo, ref, false)
+	return err
+}
+
+// createPR creates a pull request. Based on: https://godoc.org/github.com/google/go-github/github#example-PullRequestsService-Create
+func (tm triageMetadata) createPR(log shared.Logger) (err error) {
+	client := tm.githubClient
+	if prSubject == "" {
+		return errors.New("missing `-pr-title` flag; skipping PR creation")
+	}
+
+	newPR := &github.NewPullRequest{
+		Title:               &prSubject,
+		Head:                &commitBranch,
+		Base:                &prBranch,
+		Body:                &prDescription,
+		MaintainerCanModify: github.Bool(true),
+	}
+
+	pr, _, err := client.PullRequests.Create(tm.ctx, prRepoOwner, prRepo, newPR)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("PR created: %s\n", pr.GetHTMLURL())
+	return nil
+}
+
+func (tm triageMetadata) mergeToGithub(triagedMetadataMap map[string][]byte, log shared.Logger) error {
```

Comment:
This doesn't actually merge the change to GitHub, does it? It stops after creating the PR.

Actually, this has me confused a bit. As noted above, we use the same `commitBranch` for all calls to this code. Doesn't that mean calls to `triage(..)` will overlap? E.g.:

`code initializes`
- `commitBranch` set to, say, auto-triage-branch1234

1st call to `triage(..)`
- creates refs/heads/auto-triage-branch1234, pointing at master SHA `s1`
- creates a tree for the changes
- creates a new commit, `s2`, with parent `s1`, and updates refs/heads/auto-triage-branch1234 to point at `s2`
- creates a github PR from refs/heads/auto-triage-branch1234 to master

2nd call to `triage(..)`
- grabs refs/heads/auto-triage-branch1234, which points at `s2`
- creates a tree for the changes
- creates a new commit, `s3`, with parent `s2`, and updates refs/heads/auto-triage-branch1234 to point at `s3` <-- doesn't this change the *existing* PR from above?
- creates a github PR from refs/heads/auto-triage-branch1234 to master <-- but this new PR will contain *two* commits not in master?

Reply:
Related to the comment above. commitBranch is meant to be a fresh checkout of master every time so it doesn't modify any existing PR. In case of conflicts, we have to resolve merge conflicts ourselves.